File: imperitive/v2/Problem.cpp
#include "Problem.h"
#include <iostream>
#include <limits>

bool Problem::isSolved() const {

    if (combination.size() < 2) {
        return false;
    }

    set<double> criticalPrices = getCriticalPrices();
    
    for (const auto& price : criticalPrices) {
        double payoff = calculatePayoffAt(price);

        if (payoff < 0.01) {
            return false;
        }
    }
    
    return true;
}

void Problem::addLeg(const OptionLeg& leg) {
    combination.push_back(leg);
}

void Problem::removeLeg() {

    if (!combination.empty()) {
        combination.pop_back();
    }
}

void Problem::clearCombination() {
    combination.clear();
}

double Problem::getTotalCost() const {

    double total_cost = 0.0;
    
    for (const auto& leg : combination) {
        total_cost += leg.getCost();
    }
    
    return total_cost;
}

set<double> Problem::getCriticalPrices() const {

    set<double> criticalPrices;
    
    for (const auto& leg : combination) {
        criticalPrices.insert(leg.option->strike);
    }

    criticalPrices.insert(10000.0);
    criticalPrices.insert(0.0);
    
    return criticalPrices;
}

double Problem::calculatePayoffAt(double price) const {

    double payoff = -getTotalCost();
    
    for (const auto& leg : combination) {

        if (leg.option->type == OptionType::Call) {

            double intrinsic = max(0.0, price - leg.option->strike);
            payoff += (leg.position == PositionType::Long) ? intrinsic : -intrinsic;

        } else {

            double intrinsic = max(0.0, leg.option->strike - price);
            payoff += (leg.position == PositionType::Long) ? intrinsic : -intrinsic;
        }
    }
    
    return payoff;
}


vector<OptionLeg> Problem::getPossibleMoves() const {

    vector<OptionLeg> possibleMoves;

    if (combination.size() >= 4) {
        return possibleMoves; // Return empty vector
    }
    
    for (const auto& [strike, options] : market_data.strike_map) {
        // calls
        if (options.first.has_value()) {
            
            const Option* call_ptr = &options.first.value();

            // Skip adding short position for options with zero bid
            if (call_ptr->bid <= 0.00) {
                possibleMoves.push_back(OptionLeg(call_ptr, PositionType::Long));
            } else {

                bool hasLeg = false;

                for (const auto& leg : combination) {

                    if (leg.option == call_ptr) {
                        hasLeg = true;
                        break;
                    }
                }

                if(!hasLeg) {
                    possibleMoves.push_back(OptionLeg(call_ptr, PositionType::Long));
                    possibleMoves.push_back(OptionLeg(call_ptr, PositionType::Short));
                }
            }
        }
        
        // puts
        if (options.second.has_value()) {
        
            const Option* put_ptr = &options.second.value();

            // Skip adding short position for options with zero bid
            if (put_ptr->bid <= 0.0) {
                possibleMoves.push_back(OptionLeg(put_ptr, PositionType::Long));
            } else {

                bool hasLeg = false;

                for (const auto& leg : combination) {

                    if (leg.option == put_ptr) {
                        hasLeg = true;
                        break;
                    }
                }

                if(!hasLeg) {
                    possibleMoves.push_back(OptionLeg(put_ptr, PositionType::Long));
                    possibleMoves.push_back(OptionLeg(put_ptr, PositionType::Short));
                }
            }
        }
    }
    
    return possibleMoves;
}



void Problem::solve() {
    
    if (isSolved()) {
        vector<OptionLeg> sortedCombo = combination;
        sort(sortedCombo.begin(), sortedCombo.end());
        allSolutions.insert(sortedCombo);
        return;
    }
    
    
    vector<OptionLeg> possibleMoves = getPossibleMoves();
    
    for (const auto& move : possibleMoves) {
        
        addLeg(move);
        
        solve();  // Continue searching for more solutions
        
        removeLeg();
    }
}

void Problem::findAllSolutions() {
    clearSolutions();
    solve();
}

void Problem::clearSolutions() {
    allSolutions.clear();
}

void Problem::printAllSolutions() const {
    if (allSolutions.empty()) {
        cout << "\nNo arbitrage opportunities found in this option chain." << endl;
        return;
    }

    cout << "\nFound " << allSolutions.size() << " arbitrage opportunities!" << endl;

    int solutionIndex = 1;

    for (const auto& solution : allSolutions) {
        cout << "\nArbitrage Solution #" << solutionIndex++ << endl;
        cout << "--------------------------------" << endl;

        double totalCost = 0.0;
        for (const auto& leg : solution) {
            totalCost += leg.getCost();
        }

        cout << "Net Credit: $" << -totalCost << endl;
        cout << "Legs:" << endl << endl;

        int legIndex = 1;
        for (const auto& leg : solution) {
            cout << legIndex++ << ". ";
            cout << (leg.position == PositionType::Long ? "Long " : "Short ");
            cout << (leg.option->type == OptionType::Call ? "Call" : "Put");
            cout << " Strike: $" << leg.option->strike;
            cout << " Price: $" << abs(leg.getCost()) << endl;
        }

        // Calculate and print payoff
        set<double> criticalPrices;
        for (const auto& leg : solution) {
            criticalPrices.insert(leg.option->strike);
        }
        criticalPrices.insert(0.0);
        criticalPrices.insert(10000.0);

        cout << "\nPayoff at critical prices:" << endl;
        for (const auto& price : criticalPrices) {
            double payoff = -totalCost;

            for (const auto& leg : solution) {
                if (leg.option->type == OptionType::Call) {
                    double intrinsic = max(0.0, price - leg.option->strike);
                    payoff += (leg.position == PositionType::Long) ? intrinsic : -intrinsic;
                } else {
                    double intrinsic = max(0.0, leg.option->strike - price);
                    payoff += (leg.position == PositionType::Long) ? intrinsic : -intrinsic;
                }
            }

            cout << "Price: $" << price << " -> Profit: $" << payoff << endl;
        }

        cout << "--------------------------------" << endl;
    }
}

File: haskell/v3/Problem.hs
{-# LANGUAGE DeriveGeneric, DeriveAnyClass, DerivingStrategies #-}
module Problem
(
    Problem,
    chainData,
    combination,
    newProblem,
    addLeg,
    removeLeg,
    getTotalCost,
    getCriticalPrices,
    calculatePayoffAt,
    getPossibleMoves,
    isSolved,
    solve

) where

import qualified Data.Set as Set
import Data.List (sort)
import GHC.Generics (Generic)
import Control.DeepSeq (NFData)
import Option
import OptionChain
import OptionLeg

-- A Problem consists of market data (OptionChain) and a combination of legs
data Problem = ProblemCon {
    chainData :: OptionChain,
    combination :: [OptionLeg]
} deriving stock (Show, Generic)
    deriving anyclass (NFData)

-- Create a new problem with empty combination
newProblem :: OptionChain -> Problem
newProblem chain = ProblemCon chain []

-- Add leg to the combination at the head
addLeg :: OptionLeg -> Problem -> Problem
addLeg leg prob = prob {combination = leg : combination prob}

-- Remove head leg from the combination
removeLeg :: Problem -> Problem
removeLeg prob = case combination prob of
    [] -> prob
    (_:xs) -> prob {combination = xs}

-- Calculate the total cost of the combination
-- Returns cost (debit/credit)
getTotalCost :: Problem -> Double
getTotalCost prob = sum (map getCost (combination prob))

-- Get all critical prices, each unique strike and 0 and 10000
getCriticalPrices :: Problem -> Set.Set Double
getCriticalPrices prob = Set.fromList ([0.0, 10000.0] ++ [strike (legOption leg) | leg <- combination prob])

-- Calculate payoff at a specific price
-- Price -> Problem -> Payoff
calculatePayoffAt :: Double -> Problem -> Double
calculatePayoffAt price prob =
    let initialPayoff = negate (getTotalCost prob)  -- start with -cost
    in initialPayoff + sum [legPayoff leg | leg <- combination prob]
    where
        legPayoff leg =
            let opt = legOption leg
                intrinsicValue = case optionType opt of
                    Call -> max 0.0 (price - strike opt)
                    Put -> max 0.0 (strike opt - price)
            in case position leg of
                Long -> intrinsicValue
                Short -> negate intrinsicValue

-- Get all possible legs that can be added to the combination
-- Problem -> Posible moves
getPossibleMoves :: Problem -> [OptionLeg]
getPossibleMoves prob =
    if length (combination prob) >= 4
    then []  -- Maximum 4 legs
    else concat [getLegsForStrike strikePrice | strikePrice <- getAllStrikes (chainData prob)]
    where
        getLegsForStrike strikePrice =
            let (maybeCall, maybePut) = getOptionsAtStrike strikePrice (chainData prob)
                usedOptions = map legOption (combination prob)

                callLegs = case maybeCall of
                    Nothing -> []
                    Just opt ->
                        if opt `elem` usedOptions
                        then []
                        -- Skip adding short position for options with zero bid
                        else if bid opt <= 0.00
                             then [newLeg opt Long]
                             else [newLeg opt Long, newLeg opt Short]

                putLegs = case maybePut of
                    Nothing -> []
                    Just opt ->
                        if opt `elem` usedOptions
                        then []
                        -- Skip adding short position for options with zero bid
                        else if bid opt <= 0.00
                             then [newLeg opt Long]
                             else [newLeg opt Long, newLeg opt Short]
            in callLegs ++ putLegs

-- Check if the combination provides an arbitrage opportunity
isSolved :: Problem -> Bool
isSolved prob =
    let legCount = length (combination prob)
        cost = getTotalCost prob
        criticalPrices = Set.toList (getCriticalPrices prob)
        payoffs = map (`calculatePayoffAt` prob) criticalPrices
    in legCount >= 2 && all (>= 0.01) payoffs


-- Solve the problem recursively, returning all possible solutions
-- Problem -> [Solution]
solve :: Problem -> [Problem]
solve prob = removeDuplicates (findAllSolutions prob)
    where
    findAllSolutions p =
        let currentSolutions = ([p | isSolved p])
            furtherSolutions =
                if length (combination p) >= 4
                then []  -- Maximum 4 legs reached, don't explore further
                else [ solution
                     | move <- getPossibleMoves p,  -- Generate possible moves
                       let p' = addLeg move p,      -- Apply move to problem
                       solution <- findAllSolutions p'  -- Recursively solve
                     ]
        in currentSolutions ++ furtherSolutions

    -- Helper to remove duplicate solutions that just have legs in different orders
    removeDuplicates :: [Problem] -> [Problem]
    removeDuplicates [] = []
    removeDuplicates (x:xs) = x : removeDuplicates (filter (not . hasSameLegs x) xs)

    -- Check if two problems have the same set of legs (ignoring order)
    hasSameLegs :: Problem -> Problem -> Bool
    hasSameLegs p1 p2 = sort (combination p1) == sort (combination p2)